pytorch_lightning/callbacks/progress/tqdm_progress.py

```diff
@@ -105,19 +106,54 @@ def __init__(self, refresh_rate: int = 1, process_position: int = 0):
         self._refresh_rate = self._resolve_refresh_rate(refresh_rate)
         self._process_position = process_position
         self._enabled = True
-        self.main_progress_bar = None
-        self.val_progress_bar = None
-        self.test_progress_bar = None
-        self.predict_progress_bar = None
+        self._main_progress_bar: Optional[_tqdm] = None
+        self._val_progress_bar: Optional[_tqdm] = None
+        self._test_progress_bar: Optional[_tqdm] = None
+        self._predict_progress_bar: Optional[_tqdm] = None
 
-    def __getstate__(self):
+    def __getstate__(self) -> Dict:
         # can't pickle the tqdm objects
-        state = self.__dict__.copy()
-        state["main_progress_bar"] = None
-        state["val_progress_bar"] = None
-        state["test_progress_bar"] = None
-        state["predict_progress_bar"] = None
-        return state
+        return {k: v if not isinstance(v, _tqdm) else None for k, v in vars(self).items()}
+
+    @property
+    def main_progress_bar(self) -> _tqdm:
+        if self._main_progress_bar is None:
+            raise TypeError(f"The `{self.__class__.__name__}._main_progress_bar` reference has not been set yet.")
```

Comment:
I've added these properties to re-use the None check and typeerror.

The alternative was to do them everywhere they are accessed.

The callback structure of making an attribute None and creating it in a hook just doesn't work with Mypy, as it cannot know a hook will have been called.